## Keep Up Interview Challenge

### 1.) Trustworthiness of ratings

Ratings are susceptible to manipulation, bias etc. What can you say (quantitatively
speaking) about the ratings in this dataset?

For this exercise, you will analyze a dataset from Amazon. The data format and a
sample entry are shown on the next page. 

In [1]:
# load the needed packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import re

In [43]:
with open("C:\\Users\\karlk\\amazon-meta.txt", "r", encoding="utf8") as f:
    ProductPerLine = f.read().split('\n\n')

In [3]:
ProductPerLine[50000]

'Id:   49999\nASIN: B00004TGQX\n  title: Party Tyme Karaoke: Country\n  group: Music\n  salesrank: 44630\n  similar: 5  B00004TGQW  B000059T05  B00004ZBD9  B000068UZZ  B000068V01\n  categories: 6\n   |Music[5174]|Styles[301668]|Country[16]|General[63695]\n   |Music[5174]|Styles[301668]|Miscellaneous[35]|Karaoke[63945]\n   |Music[5174]|Styles[301668]|Miscellaneous[35]|General[170942]\n   |Music[5174]|Styles[301668]|Country[16]|Compilations[466016]|General[466018]\n   |Music[5174]|Specialty Stores[468040]|Indie Music[266023]|Country[171242]|General[520136]\n   |Music[5174]|Specialty Stores[468040]|Indie Music[266023]|Miscellaneous[171247]|Karaoke[560746]\n  reviews: total: 1  downloaded: 1  avg rating: 4\n    2001-1-2  cutomer: A3JN4KASA6S550  rating: 4  votes:  10  helpful:  10'

In [4]:
tot = len(ProductPerLine) -1 # first line is a description, last line is just a space

In [5]:
tot

548553

In [6]:
ProductPerLine[2]

'Id:   1\nASIN: 0827229534\n  title: Patterns of Preaching: A Sermon Sampler\n  group: Book\n  salesrank: 396585\n  similar: 5  0804215715  156101074X  0687023955  0687074231  082721619X\n  categories: 2\n   |Books[283155]|Subjects[1000]|Religion & Spirituality[22]|Christianity[12290]|Clergy[12360]|Preaching[12368]\n   |Books[283155]|Subjects[1000]|Religion & Spirituality[22]|Christianity[12290]|Clergy[12360]|Sermons[12370]\n  reviews: total: 2  downloaded: 2  avg rating: 5\n    2000-7-28  cutomer: A2JW67OY8U6HHK  rating: 5  votes:  10  helpful:   9\n    2003-12-14  cutomer: A2VE83MZF98ITY  rating: 5  votes:   6  helpful:   5'

In [7]:
# This cell takes a very long time to run, skip and read from csv after first run

# Instantiate lists for Products Table
IdStringList = []
AsinStringList = []
titleStringList = []
salesRnkStringList = []
groupStringList = []
catStringList = []

# Instantiate lists for Reviews Table
CustomerList = []
RatingList = []
VoteList = []
IdList = []
HelpfulList = []


# set up regex expressions
Id = r'Id:\s*(\d+)'
Asin = r'ASIN:\s*(\w+)'
title = r'title:\s*((.*?)\n)'
Group = r'group:\s*((.*?)\n)'
Customer = r'cutomer:\s*((.*?) )'
Rating = r'rating:\s*(\d+)'
SalesRank = r'salesrank:\s*(\d+)'
Total = r'downloaded:\s*(\d+)'
Vote = r'votes:\s*(\d+)'
Helpful = r'helpful:\s*(\d+)'
Cat = r'categories:\s*(\d+)'


# Iterate through each element to pull relevant data for products and reviews into lists
for i in range(1, tot):#tot):
    try:
        Id_temp = re.search(Id, ProductPerLine[i])
        IdStringList.append(Id_temp.group(1))
    except:
        IdStringList.append('')
    try:
        Asin_temp = re.search(Asin, ProductPerLine[i])
        AsinStringList.append(Asin_temp.group(1))
    except:
        AsinStringList.append('')
    try:
        title_temp = re.search(title, ProductPerLine[i])
        titleStringList.append(title_temp.group(1))
    except:
        titleStringList.append('')
    try:
        Group_temp = re.search(Group, ProductPerLine[i])
        groupStringList.append(Group_temp.group(1))
    except:
        groupStringList.append('')
    try:
        SaleRnk_temp = re.search(SalesRank, ProductPerLine[i])
        salesRnkStringList.append(SaleRnk_temp.group(1))
    except:
        salesRnkStringList.append('')       

    try:
        Total_temp = re.search(Total, ProductPerLine[i])
        TotalRnkString = Total_temp.group(1)
    except:
        TotalRnkString = None        

    try:
        cat_temp = re.search(Cat, ProductPerLine[2])
        catStringList.append(cat_temp.group(1))
    except:
        catStringList.append('') 


    if TotalRnkString is not None and TotalRnkString != '0':
        for j in range(0, int(TotalRnkString)):
            try:
                Customer_temp = list(re.finditer(Customer, ProductPerLine[i]))
                CustomerList.append(Customer_temp[j].group(1))
            except:
                CustomerList.append('')  

            try:
                Rating_temp = list(re.finditer(Rating, ProductPerLine[i]))
                RatingList.append(Rating_temp[j].group(1))
            except:
                RatingList.append('')  
            
            try:
                Vote_temp = list(re.finditer(Vote, ProductPerLine[i]))
                VoteList.append(Vote_temp[j].group(1))
            except:
                VoteList.append('')  
            try:
                Helpful_temp = list(re.finditer(Helpful, ProductPerLine[i]))
                HelpfulList.append(Helpful_temp[j].group(1))
            except:
                HelpfulList.append('')  
                
            IdList.append(i-1)


In [ ]:
Products = pd.DataFrame({
        'Id':  IdStringList,
        'Title': [s.strip('\n') for s in titleStringList],
        'Salesrank': salesRnkStringList,
        'Group': [s.strip('\n') for s in groupStringList],
        'CatCount': catStringList
    })
Products.head()

In [ ]:
Products.to_csv('Products.csv')

In [8]:
Reviews = pd.DataFrame({
        'Customer':  CustomerList,
        'Rating': RatingList,
        'Vote': VoteList,
        'Id': IdList, 
        'Helpful': HelpfulList
    })
Reviews.head()

Customer Helpful  Id Rating Vote
0  A2JW67OY8U6HHK        9   1      5   10
1  A2VE83MZF98ITY        5   1      5    6
2  A11NCO6YTE4BTJ        4   2      4    5
3   A9CQ3PLRNIR83        5   2      5    5
4  A13SG9ACZ9O5IM        8   2      4    8

In [12]:
Reviews.to_csv('Keep_Up_Reviews.csv')

If the data has already been cleaned then read as csv file.

In [59]:
# read csv into Products object
Products = pd.read_csv("C:/Users/karlk/Keep_Up_Products.csv", encoding = 'latin1')
# select all relevant columns, remove the row_name columns
Products = Products[['Group', 'Id', 'CatCount', 'Salesrank', 'Title']]
# Examine first rows of Products table
Products.head()

Group  Id  CatCount  Salesrank  \
0   NaN   0       NaN        NaN   
1  Book   1       2.0   396585.0   
2  Book   2       2.0   168596.0   
3  Book   3       1.0  1270652.0   
4  Book   4       5.0   631289.0   

                                               Title  
0                                                NaN  
1            Patterns of Preaching: A Sermon Sampler  
2                         Candlemas: Feast of Flames  
3   World War II Allied Fighter Planes Trading Cards  
4  Life Application Bible Commentary: 1 and 2 Tim...

In [29]:
# read csv into Products object
Reviews = pd.read_csv("C:/Users/karlk/Keep_Up_Reviews.csv", encoding = 'latin1')
# select all relevant columns, remove the row_name columns
Reviews = Reviews[['Customer', 'Helpful', 'Id', 'Rating', 'Vote']]
# Examine first rows of Products table
Reviews.head()

Customer  Helpful  Id  Rating  Vote
0  A2JW67OY8U6HHK         9   1       5    10
1  A2VE83MZF98ITY         5   1       5     6
2  A11NCO6YTE4BTJ         4   2       4     5
3   A9CQ3PLRNIR83         5   2       5     5
4  A13SG9ACZ9O5IM         8   2       4     8

In [30]:
len(Reviews)

7593244

In [31]:
# Make sure every record has a customer id
len(Reviews[Reviews['Customer'] == ''])

0

## Check if the reviews are trustworthy

Now that the data has been cleaned, lets count the number of times a given customer has reviewed a given product. 

In [32]:
# Count the number of times a given customer reviewed a product
t = Reviews[['Customer', 'Id']].groupby(['Customer', 'Id'])
t = pd.DataFrame(t.Id.count().sort_values(ascending=False))
t.columns = ['Count']

In [34]:
# Examine the highest number of reviews from a given customer for a given product
t.head()

Count
Customer        Id           
A3UN6WX5RRO2AG  534758   1193
                546219   1191
                195904   1191
                528497   1190
                507654   1188

### Find the proportion of records that have more than 1 review per customer

In [35]:
len(t[t['Count'] > 1])/len(t)

0.046069761802697264

From the cell above we can see that roughly 4.6% of the reviews have multiple reviews per customer. While some of the reviews for the same product may be for different purchases of the same product by the same customer, its hard to believe that the same customer would have bought a given product over 1,000 times (the most extreme example). This means that the reviews are not completely trustworthy.

### 2. Category bloat
Consider the product group named 'Books'. Each product in this group is associated with
categories. Naturally, with categorization, there are tradeoffs between how broad or
specific the categories must be. 

From above it was clear from the first cells that the group attribute is more or less the same as the first element in the categories though in the categories there is either a category index or a total count for a given category, most likely the later. The total count could be calculated as needed as so the first element may not be needed. See below for an example. 

In [46]:
ProductPerLine[2]

'Id:   1\nASIN: 0827229534\n  title: Patterns of Preaching: A Sermon Sampler\n  group: Book\n  salesrank: 396585\n  similar: 5  0804215715  156101074X  0687023955  0687074231  082721619X\n  categories: 2\n   |Books[283155]|Subjects[1000]|Religion & Spirituality[22]|Christianity[12290]|Clergy[12360]|Preaching[12368]\n   |Books[283155]|Subjects[1000]|Religion & Spirituality[22]|Christianity[12290]|Clergy[12360]|Sermons[12370]\n  reviews: total: 2  downloaded: 2  avg rating: 5\n    2000-7-28  cutomer: A2JW67OY8U6HHK  rating: 5  votes:  10  helpful:   9\n    2003-12-14  cutomer: A2VE83MZF98ITY  rating: 5  votes:   6  helpful:   5'

Above we can see that the first element in the category label is |Books[283155]| with minus the count is the same as the group. Also we could also potentially combine related categories such as religion, christianity, clergy, and sermons into just a religious category if we don't need the granularity. Next we will make sure that this reduces at least 10% of the data from that column.

In [60]:
# first find the total number of products that are books
len(Products[Products['Group'] == 'Book'])

393560

In [62]:
# Calculate the sum of all of the categories for books
Products[Products['Group'] == 'Book'].CatCount.sum()

1440329.0

Since each book has at least one category, Books, we can subtract the length of the book group dataframe from the sum.

In [63]:
NewCatCountTotal = Products[Products['Group'] == 'Book'].CatCount.sum()-len(Products[Products['Group'] == 'Book'])

Next Calculate the percentage difference between the sum having disgarded the books category count.

In [64]:
NewCatCountTotal/Products[Products['Group'] == 'Book'].CatCount.sum()

0.7267568729088979

Here we can see that removing the first category tag for books results in a roughly 27% decrease in category records.

### 1. Algorithm thinking
How would build the product categorization from scratch, using similar/co-purchased
information? 

To build the categorization from scratch, the similarity values could be clustered using Kmeans or a similar algorithm to group the similar items in an unsupervised manar.

### 2. Product thinking
Now, put on your 'product thinking' hat.

 1. Is it a good idea to show users the categorization hierarchy for items?
 2. Is it a good idea to show users similar/co-purchased items?
 3. Is it a good idea to show users reviews and ratings for items?
 4. For each of the above, why? How will you establish the same?

1. Showing the hierarchy for the categories is likely overkill. The similar items is probably a better solution than the categories.

2. Yes, it is a good idea to show similar/co-purchased items to the user, that's why Amazon is already doing that! The advantage is much like putting convenience items in a grocery que, people are likely to buy things impulsively if shown recommendations. Also, people might not realize that a given product is missing an iteam in order to function, like a toy that does not have batteries included. If a given user see that was a co-purchased iteam then they are more likely to realize it is not included and make the additional purchase.

3. Yes it is a good idea to show customers other users ratings (as long as they aren't biased like what was found in the trustworthy section), this can help eleviate any skepticism about a product.

4. For the similar products we could use an item based recommender solution, all other metrics would just need to be grouped, not derived.